### PR TITLE
Qdrant Quadlet contract implemented (1.2.1)

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Implemented roadmap item 1.2.1: Qdrant Podman Quadlet contract, with Rust-based validation surface, unit and behavioural tests, and documentation updates. The repository now ships the Quadlet asset, validation logic, tests, and operator guides. The roadmap is updated to reflect completion of 1.2.1.

## Changes
- New file: packaging/systemd/qdrant.container
- New Rust validation surface for Qdrant Quadlet:
  - crates/repovec-core/src/appliance/qdrant_quadlet/mod.rs
  - crates/repovec-core/src/appliance/qdrant_quadlet/error.rs
  - crates/repovec-core/src/appliance/qdrant_quadlet/parser.rs
  - crates/repovec-core/src/appliance/qdrant_quadlet/tests.rs
- New module wiring:
  - crates/repovec-core/src/appliance/mod.rs (exposes qdrant_quadlet)
- Tests and validation assets:
  - crates/repovec-core/tests/features/qdrant_quadlet.feature
  - crates/repovec-core/tests/qdrant_quadlet_bdd.rs
  - crates/repovec-core/tests/qdrant_quadlet_bdd.rs
- Documentation and planning:
  - docs/execplans/1-2-1-podman-quadlet-definition-for-qdrant.md
  - docs/repovec-appliance-technical-design.md (updated with final decisions)
  - docs/users-guide.md (new operator-facing guide for Qdrant Quadlet)
- Documentation alignment and roadmap:
  - docs/roadmap.md updated to mark 1.2.1 as done

## Rationale
- Finalizes the contract for the Qdrant Quadlet:
  - pins official image explicitly (fully qualified) and enables registry auto-update
  - binds REST and gRPC to localhost (127.0.0.1)
  - persists storage at /var/lib/repovec/qdrant-storage mounted to /qdrant/storage with SELinux relabeling
  - provides a stable Quadlet surface validated by Rust-based tooling and tests
- Keeps 1.2.1 isolated from 1.2.2/1.3.1 workstreams to avoid scope creep and enable focused validation.

## Implementation plan (summary)
- 4.1 Asset layout
  - packaging/systemd/qdrant.container checked-in as the Quadlet source
  - define install paths and Quadlet-native keys: Image, PublishPort, Volume, AutoUpdate
  - Bind ports to localhost: 127.0.0.1:6333 and 127.0.0.1:6334; storage mounted from /var/lib/repovec/qdrant-storage to /qdrant/storage
- 4.2 Quadlet shape decisions
  - Pin Qdrant image to explicit tag and keep fully qualified reference
  - SELinux relabeling via :Z and [Install] section decisions to be revisited in future scope
  - API-key wiring deferred to 1.2.2
- 4.3 Rust validation surface in repovec-core
  - Add module (e.g., crates/repovec-core/src/appliance/qdrant_quadlet.rs)
  - Implement minimal Quadlet parser and semantic validator with structured errors
- 4.4 rstest unit coverage
  - Place tests alongside validator in repovec-core with rstest fixtures
  - Cover happy path and unhappy paths (port binding, storage, auto-update, image reference, etc.)
- 4.5 rstest-bdd behavioural coverage
  - Add rstest-bdd dev dependency and BDD feature files
  - Implement step definitions for behavior validation without spinning real containers
- 4.6 Documentation updates
  - Update docs/repovec-appliance-technical-design.md with final decisions
  - Create docs/users-guide.md describing Qdrant as an appliance-internal service with loopback bindings and storage path
- 5.x Validation plan
  - Gates: fmt, markdownlint, nixie, check-fmt, lint, test
  - Manual smoke validation steps documented in the exec plan and docs

## Validation plan
- 5.1 Automated gates
  - Run: make fmt, markdownlint, nixie, check-fmt, lint, test
- 5.2 Manual smoke validation (Podman/systemd available)
  - Deploy Quadlet, start service, verify loopback bindings and storage writability, ensure reboot stability as per policy

## Documentation impact
- Updated docs/repovec-appliance-technical-design.md with final decisions (asset location, install path, pinned image policy, AutoUpdate)
- Created docs/users-guide.md describing operator-facing usage:
  - Qdrant binds only to loopback
  - Data under /var/lib/repovec/qdrant-storage
  - API-key handling deferred to 1.2.2
- Roadmap 1.2.1 marked as done once Quadlet, tests, documentation, and gates pass

## Risks and open questions
- Boot integration ownership: standalone Quadlet vs. repovec.target wiring (1.3.1)
- SELinux relabeling specifics for storage mounts
- Image pinning policy details for future iterations
- Alignment between terminal-behaviour criteria and this feature’s scope

## Definition of done
- packaging/systemd/qdrant.container exists and satisfies localhost, storage, image, and auto-update contract
- repovec-core contains Rust validation logic with rstest unit tests
- rstest-bdd behavioural tests exist for happy/unhappy paths
- Technical design and users guide reflect implemented behavior
- All gates pass: fmt, markdownlint, nixie, check-fmt, lint, test
- docs/roadmap.md updated to mark 1.2.1 as done

If you want me to adjust the level of detail or focus on specific sections (e.g., tighter testing matrix, or more rigid DoD wording), I can refine accordingly.

📎 **Task**: https://www.devboxer.com/task/c0dbc2b1-80ce-4132-9def-c53f19557ba9

### CODE REVIEW

File: crates/repovec-core/src/appliance/qdrant_quadlet/mod.rs [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+    validate_required_image(&parsed)?;
+    validate_required_port(&parsed, REQUIRED_REST_PORT, QdrantQuadletError::MissingRestPort)?;
+    validate_required_port(&parsed, REQUIRED_GRPC_PORT, QdrantQuadletError::MissingGrpcPort)?;
+    validate_storage_mount(&parsed)?;
+    validate_auto_update(&parsed)
+}
+
+fn validate_required_image(parsed: &ParsedQuadlet) -> Result<(), QdrantQuadletError> {
+    let Some(image) = parsed.values(CONTAINER_SECTION, "Image").first() else {
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/repovec-core/src/appliance/qdrant_quadlet/mod.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/repovec-core/src/appliance/qdrant_quadlet/mod.rs [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+    if !parts.get(2..).is_some_and(|options| options.contains(&"Z")) {
+        return Err(QdrantQuadletError::MissingSelinuxRelabel { volume: volume.clone() });
+    }
+
+    Ok(())
+}
+
+fn validate_auto_update(parsed: &ParsedQuadlet) -> Result<(), QdrantQuadletError> {
+    let Some(auto_update) = parsed.values(CONTAINER_SECTION, "AutoUpdate").first() else {
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/repovec-core/src/appliance/qdrant_quadlet/mod.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (2 errors, 5 warnings, 1 inconclusive)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Test suite lacks rigorous coverage of parser error cases, missing error variants, and critical integrity constraints for duplicate keys and conflicting bindings. | Add test cases exercising parser errors, duplicate/conflicting keys, and missing error variants; verify tests fail before fixes and pass after implementing validation logic. |
| Module-Level Documentation | ❌ Error | New modules lack module-level documentation comments (//!) explaining their purpose, utility and relationships within the appliance subsystem. | Add comprehensive //! documentation to each module file explaining its purpose, utility and relationships to related modules. |
| Developer Documentation | ⚠️ Warning | Developers guide lacks documentation of the new appliance module architecture, qdrant\_quadlet validation surface, and extension patterns for appliance validation. | Update docs/developers-guide.md to document appliance module architecture, validation patterns, public API surface, and test patterns used throughout the codebase. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Test suite lacks adequate unit test coverage for critical error paths and untested error variants in parser and validation logic. | Add comprehensive unit tests for parser error paths (InvalidLine, PropertyBeforeSection) and all untested validation error variants (MissingImage, UnexpectedImage, IncorrectStorageSource, MissingSelinuxRelabel). Implement test cases for duplicate Image/AutoUpdate rejection and conflicting PublishPort validation. |
| Testing (Property / Proof) | ⚠️ Warning | PR introduces parser and validator invariants requiring property-based testing to verify correct handling of arbitrary Quadlet configurations across variable whitespace, section orderings, duplicate keys, and PublishPort constraints. | Introduce property-based testing with proptest to verify parser invariants across arbitrary inputs, validate rejection of duplicate Image/AutoUpdate entries, and ensure all PublishPort entries satisfy loopback constraints. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | The PR adds 13 human-readable error messages via fmt::Display but provides zero test coverage for the formatted output; tests assert only on enum variants, never validating message text. | Add snapshot tests for all QdrantQuadletError::Display implementations using the existing snapshot infrastructure to ensure error messages remain consistent and operator-visible diagnostics stay clear. |
| Observability | ⚠️ Warning | Validation module lacks observability: no structured logging at decision points, no metrics, no tracing, no diagnostic guidance for operators. | Add structured logging (info/warn/error) at validation boundaries with contextual fields. Integrate with project's logging infrastructure and update documentation with troubleshooting guidance. |
| Domain Architecture | ❓ Inconclusive | Domain module conflates infrastructure paths (/srv/qdrant, /qdrant/storage, SELinux Z option) with domain validation contracts without clear separation of concerns. | Distinguish domain invariants (what Qdrant requires) from platform bindings (how to satisfy requirements). Extract infrastructure paths to configuration adapters or document them as enforced domain policies. |
